From 24be89330ad6998f0b8d77e566f12c843b78485e Mon Sep 17 00:00:00 2001
From: OpenEmbedded <oe.patch@oe>
Date: Wed, 12 Nov 2025 07:43:02 +0000
Subject: [PATCH 1/1] Downgrade MCTP driver and waiting for new driver
 evaluation.

Aspeed updated implementation but we might need to customize it to make it works on both Intel and AMD platforms.

Upstream-Status: Inappropriate [No upstream maintenance]
---
 drivers/net/mctp/mctp-pcie-vdm.c | 628 ++++++++++++++++++++++++++-----
 drivers/soc/aspeed/aspeed-mctp.c |  28 +-
 include/linux/aspeed-mctp.h      |   2 +-
 include/linux/mctp-pcie-vdm.h    |  17 +-
 net/mctp/route.c                 |  11 +-
 5 files changed, 575 insertions(+), 111 deletions(-)

diff --git a/drivers/net/mctp/mctp-pcie-vdm.c b/drivers/net/mctp/mctp-pcie-vdm.c
index 08ab446d24c3..56425910da9d 100644
--- a/drivers/net/mctp/mctp-pcie-vdm.c
+++ b/drivers/net/mctp/mctp-pcie-vdm.c
@@ -37,6 +37,18 @@
 #define MCTP_PCIE_VDM_MAX_MTU 512
 /* 16byte */
 #define MCTP_PCIE_VDM_HDR_SIZE 16
+#define MCTP_PCIE_VDM_HDR_DEST_EID_OFFSET 13
+#define MCTP_PCIE_VDM_HDR_SRC_EID_OFFSET 14
+#define MCTP_PAYLOAD_IC_TYPE_SIZE 1
+#define MCTP_RECEIVE_PKT_TIMEOUT_MS 5
+
+#define MCTP_PCIE_VDM_NET_DEV_TX_QUEUE_LEN 1100
+#define MCTP_PCIE_VDM_DEV_TX_QUEUE_SIZE 64
+
+#define MCTP_PCIE_VDM_FMT_4DW 0x3
+#define MCTP_PCIE_VDM_TYPE_MSG 0x10
+#define MCTP_PCIE_VDM_CODE 0x0
+/* PCIe VDM message code */
 #define MCTP_PAYLOAD_IC_TYPE_SIZE 1
 #define MCTP_RECEIVE_PKT_TIMEOUT_MS 5
 
@@ -52,6 +64,13 @@
 /* MCTP message type */
 #define MCTP_MSG_TYPE_MASK GENMASK(6, 0)
 #define MCTP_PCIE_VDM_MSG_TYPE 0x7E
+#define MCTP_CONTROL_MSG_TYPE 0x00
+
+#define MCTP_CTRL_MSG_RQDI_REQ 0x80
+#define MCTP_CTRL_MSG_RQDI_RSP 0x00
+
+#define MCTP_CTRL_MSG_RQDI_REQ_DATA_OFFSET 3
+#define MCTP_CTRL_MSG_RQDI_RSP_DATA_OFFSET 4
 
 #define MCTP_PCIE_SWAP_NET_ENDIAN(arr, len)       \
 	do {                                      \
@@ -84,6 +103,11 @@ enum mctp_ctrl_command_code {
 	MCTP_CTRL_CMD_DISCOVERY_NOTIFY = 0x0D
 };
 
+struct mctp_ctrl_msg_hdr {
+	u8 instanceID : 5, rsvd : 1, datagram : 1, rq : 1;
+	u8 command_code;
+};
+
 struct mctp_pcie_vdm_hdr {
 	u32 length : 10, rsvd0 : 2, attr : 2, ep : 1, td : 1, rsvd1 : 4, tc : 3,
 		rsvd2 : 1, route_type : 5, fmt : 2, rsvd3 : 1;
@@ -94,11 +118,30 @@ struct mctp_pcie_vdm_hdr {
 	u16 pci_target_id;
 };
 
+struct mctp_pcie_vdm_route_info {
+	u8 eid;
+	u8 dirty;
+	u16 bdf_addr;
+	struct hlist_node hnode;
+};
+
 struct mctp_pcie_vdm_dev {
 	struct device *dev;
+	struct net_device *ndev;
+	struct task_struct *tx_thread;
+	wait_queue_head_t tx_wait;
+	struct work_struct rx_work;
+	struct ptr_ring tx_queue;
+	struct list_head list;
+	/* each network may have at most 256 EIDs */
+	DECLARE_HASHTABLE(route_table, 8);
 	const struct mctp_pcie_vdm_ops *callback_ops;
 };
 
+/* mutex for vdm_devs add/delete */
+DEFINE_MUTEX(mctp_pcie_vdm_dev_mutex);
+LIST_HEAD(mctp_pcie_vdm_devs);
+struct workqueue_struct *mctp_pcie_vdm_wq;
 
 static const struct mctp_pcie_vdm_hdr mctp_pcie_vdm_hdr_template = {
 	.fmt = MCTP_PCIE_VDM_FMT_4DW,
@@ -130,22 +173,300 @@ static void mctp_pcie_vdm_display_skb_buff_data(struct sk_buff *skb)
 	pr_debug("%s\n", buf);
 }
 
-static int mctp_pcie_vdm_xmit(struct net_device *ndev, struct sk_buff *skb)
+static void mctp_pcie_vdm_update_route_table(struct mctp_pcie_vdm_dev *vdm_dev,
+					     u8 eid, u16 bdf)
+{
+	if (eid == 0x00 || eid == 0xFF)
+		return;
+
+	bool exist = false;
+	struct mctp_pcie_vdm_route_info *route;
+
+	hash_for_each_possible(vdm_dev->route_table, route, hnode, eid) {
+		pr_debug("%s: route table eid %d maps to %d", __func__,
+			 route->eid, route->bdf_addr);
+		if (route->eid == eid) {
+			exist = true;
+			route->bdf_addr = bdf;
+			break;
+		}
+	}
+
+	if (!exist) {
+		route = kmalloc(sizeof(*route), GFP_KERNEL);
+		route->bdf_addr = bdf;
+		route->eid = eid;
+		route->dirty = 0;
+
+		hash_add(vdm_dev->route_table, &route->hnode, route->eid);
+		pr_debug("%s: not found, add map eid %d to bdf 0x%x", __func__,
+			 eid, bdf);
+	}
+}
+
+static void mctp_pcie_vdm_ctrl_msg_handler(struct mctp_pcie_vdm_dev *vdm_dev,
+					   u8 *packet, bool isTx)
+{
+	u8 message_type =
+		FIELD_GET(MCTP_MSG_TYPE_MASK, packet[MCTP_PCIE_VDM_HDR_SIZE]);
+
+	pr_debug("%s: Message Type: %d", __func__, message_type);
+
+	if (message_type != MCTP_CONTROL_MSG_TYPE)
+		return;
+
+	struct mctp_ctrl_msg_hdr *ctrl_hdr =
+		(struct mctp_ctrl_msg_hdr *)(&packet[MCTP_PCIE_VDM_HDR_SIZE + 1]);
+
+	/* host endian expected */
+	struct mctp_pcie_vdm_hdr *hdr = (struct mctp_pcie_vdm_hdr *)packet;
+
+	pr_debug("%s: Command Code: %d", __func__, ctrl_hdr->command_code);
+    static u8 bo_eid = 0;
+    static u8 rc_eid = 0;
+
+    if (isTx) {
+        pr_debug("%s: TX (%d) --> (%d)", __func__, packet[MCTP_PCIE_VDM_HDR_SRC_EID_OFFSET], packet[MCTP_PCIE_VDM_HDR_DEST_EID_OFFSET]);
+        // BMC is Requester
+        switch (ctrl_hdr->command_code) {
+        case MCTP_CTRL_CMD_SET_ENDPOINT_ID:
+            if ((ctrl_hdr->rq == 1) && (ctrl_hdr->datagram == 0)) {
+                pr_debug("%s: MCTP_CTRL_CMD_SET_ENDPOINT_ID from BMC", __func__);
+                /* EID placed at byte2 of SET EID REQ DATA */
+                bo_eid = packet[MCTP_PCIE_VDM_HDR_SRC_EID_OFFSET];
+
+                u8 dst_eid =
+                    packet[MCTP_PCIE_VDM_HDR_SIZE +
+                           MCTP_CTRL_MSG_RQDI_REQ_DATA_OFFSET + 1];
+                u16 dst_bdf = hdr->pci_target_id;
+
+                mctp_pcie_vdm_update_route_table(vdm_dev, dst_eid,
+                                 dst_bdf);
+                packet[MCTP_PCIE_VDM_HDR_DEST_EID_OFFSET] = 0xFF;
+                hdr->route_type = MCTP_PCIE_VDM_TYPE_MSG |
+                          MCTP_PCIE_VDM_BROADCAST_FROM_RC;
+                hdr->pci_target_id = 0xFFFF;
+            } else if ((ctrl_hdr->rq == 0) && (ctrl_hdr->datagram == 0)) {
+                packet[MCTP_PCIE_VDM_HDR_DEST_EID_OFFSET] = bo_eid;
+                u16 dst_bdf = hdr->pci_target_id;
+                mctp_pcie_vdm_update_route_table(vdm_dev, bo_eid,
+                                 dst_bdf);
+            }
+            break;
+        case MCTP_CTRL_CMD_GET_ENDPOINT_ID:
+            if ((ctrl_hdr->rq == 0) && (ctrl_hdr->datagram == 0)) {
+                /* EID placed at byte2 of GET EID RSP DATA */
+                u8 target_eid =
+                    packet[MCTP_PCIE_VDM_HDR_SIZE +
+                           MCTP_CTRL_MSG_RQDI_RSP_DATA_OFFSET + 1];
+                u16 src_bdf = hdr->pci_req_id;
+
+                mctp_pcie_vdm_update_route_table(vdm_dev, target_eid,
+                                 src_bdf);
+            }
+            break;
+        case MCTP_CTRL_CMD_DISCOVERY_NOTIFY:
+            hdr->pci_target_id = 0x0000;
+            /* default use MCTP_PCIE_VDM_ROUTE_BY_ID, so no need to handle RSP class */
+            if ((ctrl_hdr->rq == 1) && (ctrl_hdr->datagram == 0)) {
+                hdr->route_type = MCTP_PCIE_VDM_TYPE_MSG |
+                          MCTP_PCIE_VDM_ROUTE_TO_RC;
+            }
+            break;
+        case MCTP_CTRL_CMD_PREPARE_ENDPOINT_DISCOVERY:
+        case MCTP_CTRL_CMD_ENDPOINT_DISCOVERY:
+            if ((ctrl_hdr->rq == 1) && (ctrl_hdr->datagram == 0)) {
+                bo_eid = packet[MCTP_PCIE_VDM_HDR_SRC_EID_OFFSET];
+                rc_eid = packet[MCTP_PCIE_VDM_HDR_DEST_EID_OFFSET];
+
+                pr_debug("%s: MCTP_CTRL_CMD_ENDPOINT_DISCOVERY Req from %02x: %x", __func__,
+                    hdr->pci_req_id,
+                    hdr->route_type);
+                packet[MCTP_PCIE_VDM_HDR_DEST_EID_OFFSET] = 0xFF;
+                hdr->route_type = MCTP_PCIE_VDM_TYPE_MSG |
+                          MCTP_PCIE_VDM_BROADCAST_FROM_RC;
+                hdr->pci_target_id = 0xFFFF;
+                u16 src_bdf = hdr->pci_req_id;
+            } else if ((ctrl_hdr->rq == 0) && (ctrl_hdr->datagram == 0)) {
+                packet[MCTP_PCIE_VDM_HDR_SRC_EID_OFFSET] = rc_eid;
+                pr_debug("%s: MCTP_CTRL_CMD_ENDPOINT_DISCOVERY Resp: %x", __func__,
+                    hdr->route_type);
+            }
+            break;
+        default:
+            /* Unknown command code or not supported currently */
+            pr_debug("%s: Unsupported Command Code: %d", __func__,
+                 ctrl_hdr->command_code);
+            break;
+        }
+
+        if ((ctrl_hdr->rq == 0) && (ctrl_hdr->datagram == 0) && (packet[MCTP_PCIE_VDM_HDR_DEST_EID_OFFSET] == 0)) {
+            pr_debug("%s: Target is NULL in response message. Patched EID to %d", __func__, bo_eid);
+            packet[MCTP_PCIE_VDM_HDR_DEST_EID_OFFSET] = bo_eid;
+        }
+    } else {
+        // BMC is Receiver.
+        pr_debug("%s: RX (%d) --> (%d)", __func__, packet[MCTP_PCIE_VDM_HDR_SRC_EID_OFFSET], packet[MCTP_PCIE_VDM_HDR_DEST_EID_OFFSET]);
+        switch (ctrl_hdr->command_code) {
+        case MCTP_CTRL_CMD_SET_ENDPOINT_ID:
+            if ((ctrl_hdr->rq == 1) && (ctrl_hdr->datagram == 0)) {
+                pr_debug("%s: MCTP_CTRL_CMD_SET_ENDPOINT_ID from RC", __func__);
+                /* EID placed at byte2 of SET EID REQ DATA */
+                bo_eid = packet[MCTP_PCIE_VDM_HDR_SRC_EID_OFFSET];
+
+                u8 dst_eid =
+                    packet[MCTP_PCIE_VDM_HDR_SIZE +
+                           MCTP_CTRL_MSG_RQDI_REQ_DATA_OFFSET + 1];
+                u16 dst_bdf = hdr->pci_target_id;
+
+                mctp_pcie_vdm_update_route_table(vdm_dev, dst_eid,
+                                 dst_bdf);
+                packet[MCTP_PCIE_VDM_HDR_DEST_EID_OFFSET] = 0xFF;
+                hdr->route_type = MCTP_PCIE_VDM_TYPE_MSG |
+                          MCTP_PCIE_VDM_BROADCAST_FROM_RC;
+                hdr->pci_target_id = 0xFFFF;
+            } else if ((ctrl_hdr->rq == 0) && (ctrl_hdr->datagram == 0)) {
+                packet[MCTP_PCIE_VDM_HDR_DEST_EID_OFFSET] = bo_eid;
+                u16 dst_bdf = hdr->pci_target_id;
+                mctp_pcie_vdm_update_route_table(vdm_dev, bo_eid,
+                                 dst_bdf);
+            }
+            break;
+        case MCTP_CTRL_CMD_GET_ENDPOINT_ID:
+            if ((ctrl_hdr->rq == 0) && (ctrl_hdr->datagram == 0)) {
+                /* EID placed at byte2 of GET EID RSP DATA */
+                u8 target_eid =
+                    packet[MCTP_PCIE_VDM_HDR_SIZE +
+                           MCTP_CTRL_MSG_RQDI_RSP_DATA_OFFSET + 1];
+                u16 src_bdf = hdr->pci_req_id;
+
+                mctp_pcie_vdm_update_route_table(vdm_dev, target_eid,
+                                 src_bdf);
+            }
+            break;
+        case MCTP_CTRL_CMD_DISCOVERY_NOTIFY:
+            hdr->pci_target_id = 0x0000;
+            /* default use MCTP_PCIE_VDM_ROUTE_BY_ID, so no need to handle RSP class */
+            if ((ctrl_hdr->rq == 1) && (ctrl_hdr->datagram == 0)) {
+                hdr->route_type = MCTP_PCIE_VDM_TYPE_MSG |
+                          MCTP_PCIE_VDM_ROUTE_TO_RC;
+            }
+            break;
+        case MCTP_CTRL_CMD_PREPARE_ENDPOINT_DISCOVERY:
+        case MCTP_CTRL_CMD_ENDPOINT_DISCOVERY:
+            if ((ctrl_hdr->rq == 1) && (ctrl_hdr->datagram == 0)) {
+                pr_debug("%s: MCTP_CTRL_CMD_ENDPOINT_DISCOVERY Req from RC %02x: %x", __func__,
+                    hdr->pci_req_id,
+                    hdr->route_type);
+                hdr->route_type = MCTP_PCIE_VDM_TYPE_MSG |
+                          MCTP_PCIE_VDM_BROADCAST_FROM_RC;
+                hdr->pci_target_id = 0xFFFF;
+                u16 src_bdf = hdr->pci_req_id;
+            } else if ((ctrl_hdr->rq == 0) && (ctrl_hdr->datagram == 0)) {
+                hdr->route_type = MCTP_PCIE_VDM_TYPE_MSG |
+                          MCTP_PCIE_VDM_ROUTE_TO_RC;
+                pr_debug("%s: MCTP_CTRL_CMD_ENDPOINT_DISCOVERY Resp from RC: %x", __func__,
+                    hdr->route_type);
+            }
+            break;
+        default:
+            /* Unknown command code or not supported currently */
+            pr_debug("%s: Unsupported Command Code: %d", __func__,
+                 ctrl_hdr->command_code);
+            break;
+        }
+
+        // Received an response
+        if ((ctrl_hdr->rq == 0) && (ctrl_hdr->datagram == 0)) {
+            if (packet[MCTP_PCIE_VDM_HDR_DEST_EID_OFFSET] == 0) {
+                pr_debug("%s: Target is NULL in response message. Patched EID to %d", __func__, bo_eid);
+                packet[MCTP_PCIE_VDM_HDR_DEST_EID_OFFSET] = bo_eid;
+            }
+            if (packet[MCTP_PCIE_VDM_HDR_SRC_EID_OFFSET] == 0) {
+                pr_debug("%s: Source is NULL in response message. Patched EID to %d", __func__, rc_eid);
+                packet[MCTP_PCIE_VDM_HDR_SRC_EID_OFFSET] = rc_eid;
+            }
+        }
+    }
+}
+
+static netdev_tx_t mctp_pcie_vdm_start_xmit(struct sk_buff *skb,
+					    struct net_device *ndev)
+{
+	struct mctp_pcie_vdm_dev *vdm_dev = netdev_priv(ndev);
+
+	pr_debug("%s: skb len %u\n", __func__, skb->len);
+
+	netdev_tx_t ret;
+
+	if (ptr_ring_full(&vdm_dev->tx_queue)) {
+		pr_debug("%s: failed to send packet, buffer full\n", __func__);
+		netif_stop_queue(ndev);
+
+		ret = NETDEV_TX_BUSY;
+	} else {
+		int reason = ptr_ring_produce_bh(&vdm_dev->tx_queue, skb);
+
+		if (reason) {
+			pr_err("%s: failed to produce skb to tx queue, reason %d\n",
+			       __func__, reason);
+			ret = NETDEV_TX_BUSY;
+		} else {
+			ret = NETDEV_TX_OK;
+		}
+	}
+
+	wake_up(&vdm_dev->tx_wait);
+
+	return ret;
+}
+
+static void mctp_pcie_vdm_xmit(struct mctp_pcie_vdm_dev *vdm_dev,
+			       struct sk_buff *skb)
 {
 	struct net_device_stats *stats;
 	struct mctp_pcie_vdm_hdr *hdr;
-	struct mctp_pcie_vdm_dev *vdm_dev;
 	u8 *hdr_byte;
+	u8 message_type;
 	u16 payload_len_dw;
 	u16 payload_len_byte;
 	int rc;
 
-	stats = &ndev->stats;
-	vdm_dev = netdev_priv(ndev);
+	stats = &vdm_dev->ndev->stats;
 	hdr = (struct mctp_pcie_vdm_hdr *)skb->data;
 	hdr_byte = skb->data;
 	payload_len_dw = (ALIGN(skb->len, sizeof(u32)) - MCTP_PCIE_VDM_HDR_SIZE) / sizeof(u32);
 	payload_len_byte = skb->len - MCTP_PCIE_VDM_HDR_SIZE;
+	message_type = FIELD_GET(MCTP_MSG_TYPE_MASK, hdr_byte[MCTP_PCIE_VDM_HDR_SIZE]);
+
+	if (message_type == MCTP_CONTROL_MSG_TYPE) {
+		mctp_pcie_vdm_ctrl_msg_handler(vdm_dev, hdr_byte, true);
+	} else {
+		if (hdr->route_type ==
+		    (MCTP_PCIE_VDM_TYPE_MSG | MCTP_PCIE_VDM_ROUTE_BY_ID)) {
+			struct mctp_pcie_vdm_route_info *route;
+			bool exist = false;
+			u16 bdf = 0x00;
+			u8 dst_eid =
+				FIELD_GET(GENMASK(15, 8),
+					  *((u32 *)skb->data +
+					    sizeof(struct mctp_pcie_vdm_hdr) /
+						    sizeof(u32)));
+
+			hash_for_each_possible(vdm_dev->route_table, route,
+					       hnode, dst_eid) {
+				if (route->eid == dst_eid) {
+					exist = true;
+					bdf = route->bdf_addr;
+					hdr->pci_target_id = bdf;
+					break;
+				}
+			}
+			if (exist)
+				pr_debug("%s fill bdf 0x%x to eid %d", __func__,
+					 bdf, dst_eid);
+		}
+	}
 
 	hdr->length = payload_len_dw;
 	hdr->tag_pad_len =
@@ -161,42 +482,177 @@ static int mctp_pcie_vdm_xmit(struct net_device *ndev, struct sk_buff *skb)
 	if (rc) {
 		pr_err("%s: failed to send packet, rc %d\n", __func__, rc);
 		stats->tx_errors++;
-	} else {
-		stats->tx_packets++;
-		stats->tx_bytes += (skb->len - sizeof(struct mctp_pcie_vdm_hdr));
+		return;
 	}
-	return rc;
+	stats->tx_packets++;
+	stats->tx_bytes += (skb->len - sizeof(struct mctp_pcie_vdm_hdr));
 }
 
-static netdev_tx_t mctp_pcie_vdm_start_xmit(struct sk_buff *skb,
-					    struct net_device *ndev)
+static int mctp_pcie_vdm_tx_thread(void *data)
 {
-	int rc;
-	netdev_tx_t ret;
+	struct mctp_pcie_vdm_dev *vdm_dev = data;
 
-	pr_debug("%s: skb len %u\n", __func__, skb->len);
+	for (;;) {
+		wait_event_idle(vdm_dev->tx_wait,
+				__ptr_ring_peek(&vdm_dev->tx_queue) ||
+					kthread_should_stop());
 
-	if (skb) {
-		rc = mctp_pcie_vdm_xmit(ndev, skb);
-		if (rc) {
-			pr_err("%s: failed to send packet, rc %d\n", __func__, rc);
-			ret = NETDEV_TX_BUSY;
+		if (kthread_should_stop())
+			break;
+
+		while (__ptr_ring_peek(&vdm_dev->tx_queue)) {
+			struct sk_buff *skb;
+
+			skb = ptr_ring_consume(&vdm_dev->tx_queue);
+
+			if (skb) {
+				mctp_pcie_vdm_xmit(vdm_dev, skb);
+				kfree_skb(skb);
+			}
+		}
+		if (netif_queue_stopped(vdm_dev->ndev))
+			netif_wake_queue(vdm_dev->ndev);
+	}
+
+	pr_debug("%s stopping\n", __func__);
+	return 0;
+}
+
+static void mctp_pcie_vdm_rx_work_handler(struct work_struct *work)
+{
+	struct mctp_pcie_vdm_dev *vdm_dev;
+	u8 *packet;
+
+	vdm_dev = container_of(work, struct mctp_pcie_vdm_dev, rx_work);
+	packet = vdm_dev->callback_ops->recv_packet(vdm_dev->dev);
+
+	while (!IS_ERR(packet)) {
+		MCTP_PCIE_SWAP_LITTLE_ENDIAN((u32 *)packet,
+					     sizeof(struct mctp_pcie_vdm_hdr) / sizeof(u32));
+		struct mctp_pcie_vdm_hdr *vdm_hdr = (struct mctp_pcie_vdm_hdr *)packet;
+		struct mctp_skb_cb *cb;
+		struct net_device_stats *stats;
+		struct sk_buff *skb;
+		u16 len;
+		int net_status;
+
+		stats = &vdm_dev->ndev->stats;
+		len = vdm_hdr->length * sizeof(u32) -
+				vdm_hdr->tag_pad_len;
+		len += (MCTP_PCIE_VDM_HDR_SIZE - sizeof(struct mctp_pcie_vdm_hdr));
+		skb = netdev_alloc_skb(vdm_dev->ndev, len);
+		pr_debug("%s: received packet size: %d\n", __func__,
+			 len);
+
+		if (!skb) {
+			stats->rx_errors++;
+			pr_err("%s: failed to alloc skb\n", __func__);
+			continue;
+		}
+
+		skb->protocol = htons(ETH_P_MCTP);
+		/* put data into tail sk buff */
+        // Patch received data first
+		mctp_pcie_vdm_ctrl_msg_handler(vdm_dev, packet, false);
+		skb_put_data(skb, &packet[sizeof(struct mctp_pcie_vdm_hdr)], len);
+		mctp_pcie_vdm_display_skb_buff_data(skb);
+
+		cb = __mctp_cb(skb);
+		cb->halen = 2; // BDF size is 2 bytes
+		memcpy(cb->haddr, &vdm_hdr->pci_req_id, cb->halen);
+
+		net_status = netif_rx(skb);
+		if (net_status == NET_RX_SUCCESS) {
+			stats->rx_packets++;
+			stats->rx_bytes += len;
 		} else {
-			ret = NETDEV_TX_OK;
-			kfree_skb(skb);
+			stats->rx_dropped++;
+		}
+
+		//mctp_pcie_vdm_ctrl_msg_handler(vdm_dev, packet, false);
+
+		if (vdm_hdr->route_type ==
+			(MCTP_PCIE_VDM_TYPE_MSG |
+				MCTP_PCIE_VDM_ROUTE_BY_ID)) {
+			u32 mctp_hdr;
+			u16 bdf;
+			u8 src_eid;
+
+			bdf = vdm_hdr->pci_req_id;
+			mctp_hdr = *((u32 *)packet +
+				     sizeof(struct mctp_pcie_vdm_hdr) /
+					     sizeof(u32));
+			MCTP_PCIE_SWAP_LITTLE_ENDIAN(&mctp_hdr, 1);
+			src_eid = FIELD_GET(GENMASK(15, 8), mctp_hdr);
+
+			mctp_pcie_vdm_update_route_table(vdm_dev,
+							 src_eid, bdf);
 		}
+
+		vdm_dev->callback_ops->free_packet(packet);
+		packet = vdm_dev->callback_ops->recv_packet(vdm_dev->dev);
 	}
-	return ret;
+}
+
+static int mctp_pcie_vdm_add_mctp_dev(struct mctp_pcie_vdm_dev *vdm_dev)
+{
+	hash_init(vdm_dev->route_table);
+	INIT_LIST_HEAD(&vdm_dev->list);
+	init_waitqueue_head(&vdm_dev->tx_wait);
+	ptr_ring_init(&vdm_dev->tx_queue, MCTP_PCIE_VDM_DEV_TX_QUEUE_SIZE,
+		      GFP_KERNEL);
+	vdm_dev->tx_thread = kthread_run(mctp_pcie_vdm_tx_thread, vdm_dev,
+					 "mctp_pcie_vdm_tx_thread");
+	INIT_WORK(&vdm_dev->rx_work, mctp_pcie_vdm_rx_work_handler);
+
+	mutex_lock(&mctp_pcie_vdm_dev_mutex);
+	list_add_tail(&vdm_dev->list, &mctp_pcie_vdm_devs);
+	mutex_unlock(&mctp_pcie_vdm_dev_mutex);
+	return 0;
 }
 
 static void mctp_pcie_vdm_uninit(struct net_device *ndev)
 {
 	struct mctp_pcie_vdm_dev *vdm_dev;
 
+	struct mctp_pcie_vdm_route_info *route;
+	struct hlist_node *tmp;
+	int bkt;
+
 	vdm_dev = netdev_priv(ndev);
 	pr_info("%s: uninitializing vdm_dev %s\n", __func__,
-		ndev->name);
+		vdm_dev->ndev->name);
 	vdm_dev->callback_ops->uninit(vdm_dev->dev);
+
+	hash_for_each_safe(vdm_dev->route_table, bkt, tmp, route, hnode) {
+		hash_del(&route->hnode);
+		kfree(route);
+	}
+
+	if (vdm_dev->tx_thread) {
+		kthread_stop(vdm_dev->tx_thread);
+		vdm_dev->tx_thread = NULL;
+	}
+
+	if (mctp_pcie_vdm_wq) {
+		cancel_work_sync(&vdm_dev->rx_work);
+		flush_workqueue(mctp_pcie_vdm_wq);
+		destroy_workqueue(mctp_pcie_vdm_wq);
+		mctp_pcie_vdm_wq = NULL;
+	}
+
+	while (__ptr_ring_peek(&vdm_dev->tx_queue)) {
+		struct sk_buff *skb;
+
+		skb = ptr_ring_consume(&vdm_dev->tx_queue);
+
+		if (skb)
+			kfree_skb(skb);
+	}
+
+	mutex_lock(&mctp_pcie_vdm_dev_mutex);
+	list_del(&vdm_dev->list);
+	mutex_unlock(&mctp_pcie_vdm_dev_mutex);
 }
 
 static int mctp_pcie_vdm_hdr_create(struct sk_buff *skb,
@@ -204,17 +660,14 @@ static int mctp_pcie_vdm_hdr_create(struct sk_buff *skb,
 				    unsigned short type, const void *daddr,
 				    const void *saddr, unsigned int len)
 {
-	u8 dest_addr[3] = {0};
 	struct mctp_pcie_vdm_hdr *hdr =
 		(struct mctp_pcie_vdm_hdr *)skb_push(skb, sizeof(*hdr));
 
 	pr_debug("%s type %d len %d\n", __func__, type, len);
 	memcpy(hdr, &mctp_pcie_vdm_hdr_template, sizeof(*hdr));
 	if (daddr) {
-		memcpy(dest_addr, (u8 *)daddr, sizeof(dest_addr));
-		hdr->route_type |= dest_addr[0] & GENMASK(2, 0);
-		hdr->pci_target_id = dest_addr[1] << 8 | dest_addr[2];
-		pr_debug("%s dst route %d addr %d\n", __func__, hdr->route_type, hdr->pci_target_id);
+		pr_debug("%s dst addr %d\n", __func__, *(u16 *)daddr);
+		hdr->pci_target_id = *(u16 *)daddr;
 	}
 
 	if (saddr) {
@@ -242,7 +695,7 @@ static void mctp_pcie_vdm_net_setup(struct net_device *ndev)
 	ndev->min_mtu = MCTP_PCIE_VDM_MIN_MTU;
 	ndev->max_mtu = MCTP_PCIE_VDM_MAX_MTU;
 	ndev->tx_queue_len = MCTP_PCIE_VDM_NET_DEV_TX_QUEUE_LEN;
-	ndev->addr_len = 3; //PCIe bdf is 2bytes + 1byte route type
+	ndev->addr_len = 2; //PCIe bdf is 2 bytes
 	ndev->hard_header_len = sizeof(struct mctp_pcie_vdm_hdr);
 
 	ndev->netdev_ops = &mctp_pcie_vdm_net_ops;
@@ -273,67 +726,10 @@ static int mctp_pcie_vdm_add_net_dev(struct net_device **dev)
 	return rc;
 }
 
-void mctp_pcie_vdm_receive_packet(struct net_device *ndev)
-{
-	struct mctp_pcie_vdm_dev *vdm_dev;
-	u8 *packet;
-
-	vdm_dev = netdev_priv(ndev);
-	packet = vdm_dev->callback_ops->recv_packet(vdm_dev->dev);
-
-	while (!IS_ERR(packet)) {
-		MCTP_PCIE_SWAP_LITTLE_ENDIAN((u32 *)packet,
-					     sizeof(struct mctp_pcie_vdm_hdr) / sizeof(u32));
-		struct mctp_pcie_vdm_hdr *vdm_hdr = (struct mctp_pcie_vdm_hdr *)packet;
-		struct mctp_skb_cb *cb;
-		struct net_device_stats *stats;
-		struct sk_buff *skb;
-		u16 len;
-		int net_status;
-
-		stats = &ndev->stats;
-		len = vdm_hdr->length * sizeof(u32) -
-				vdm_hdr->tag_pad_len;
-		len += (MCTP_PCIE_VDM_HDR_SIZE - sizeof(struct mctp_pcie_vdm_hdr));
-		skb = netdev_alloc_skb(ndev, len);
-		pr_debug("%s: received packet size: %d\n", __func__,
-			 len);
-
-		if (!skb) {
-			stats->rx_errors++;
-			pr_err("%s: failed to alloc skb\n", __func__);
-			continue;
-		}
-
-		skb->protocol = htons(ETH_P_MCTP);
-		/* put data into tail sk buff */
-		skb_put_data(skb, &packet[sizeof(struct mctp_pcie_vdm_hdr)], len);
-		mctp_pcie_vdm_display_skb_buff_data(skb);
-
-		cb = __mctp_cb(skb);
-		cb->halen = 3; // route type | bdf address
-		cb->haddr[0] = vdm_hdr->route_type;
-		cb->haddr[1] = vdm_hdr->pci_req_id >> 8;
-		cb->haddr[2] = vdm_hdr->pci_req_id & 0xFF;
-
-		net_status = netif_rx(skb);
-		if (net_status == NET_RX_SUCCESS) {
-			stats->rx_packets++;
-			stats->rx_bytes += len;
-		} else {
-			stats->rx_dropped++;
-		}
-
-		vdm_dev->callback_ops->free_packet(packet);
-		packet = vdm_dev->callback_ops->recv_packet(vdm_dev->dev);
-	}
-}
-
-struct net_device *mctp_pcie_vdm_add_dev(struct device *dev,
-					 const struct mctp_pcie_vdm_ops *ops)
+struct mctp_pcie_vdm_dev *mctp_pcie_vdm_add_dev(struct device *dev,
+						const struct mctp_pcie_vdm_ops *ops)
 {
 	struct net_device *ndev;
-	struct mctp_pcie_vdm_dev *vdm_dev;
 	int rc;
 
 	rc = mctp_pcie_vdm_add_net_dev(&ndev);
@@ -342,21 +738,67 @@ struct net_device *mctp_pcie_vdm_add_dev(struct device *dev,
 		return ERR_PTR(rc);
 	}
 
+	struct mctp_pcie_vdm_dev *vdm_dev;
+
 	vdm_dev = netdev_priv(ndev);
+	vdm_dev->ndev = ndev;
 	vdm_dev->dev = dev;
 	vdm_dev->callback_ops = ops;
 
-	return ndev;
+	rc = mctp_pcie_vdm_add_mctp_dev(vdm_dev);
+	if (rc) {
+		pr_err("%s: failed to add mctp device\n", __func__);
+		unregister_netdev(ndev);
+		free_netdev(ndev);
+		return ERR_PTR(rc);
+	}
+	return vdm_dev;
 }
 EXPORT_SYMBOL_GPL(mctp_pcie_vdm_add_dev);
 
-void mctp_pcie_vdm_remove_dev(struct net_device *vdm_dev)
+void mctp_pcie_vdm_remove_dev(struct mctp_pcie_vdm_dev *vdm_dev)
 {
-	pr_debug("%s: removing vdm_dev %s\n", __func__, vdm_dev->name);
+	pr_debug("%s: removing vdm_dev %s\n", __func__, vdm_dev->ndev->name);
+	struct net_device *ndev = vdm_dev->ndev;
 
-	if (vdm_dev) {
-		mctp_unregister_netdev(vdm_dev);
-		free_netdev(vdm_dev);
+	if (ndev) {
+		// objects in vdm_dev will be released in net_dev uninit operator
+		mctp_unregister_netdev(ndev);
+		free_netdev(ndev);
 	}
 }
 EXPORT_SYMBOL_GPL(mctp_pcie_vdm_remove_dev);
+
+void mctp_pcie_vdm_notify_rx(struct mctp_pcie_vdm_dev *vdm_dev)
+{
+	queue_work(mctp_pcie_vdm_wq, &vdm_dev->rx_work);
+}
+EXPORT_SYMBOL_GPL(mctp_pcie_vdm_notify_rx);
+
+static __init int mctp_pcie_vdm_mod_init(void)
+{
+	mctp_pcie_vdm_wq = alloc_workqueue("mctp_pcie_vdm_wq", WQ_UNBOUND, 1);
+	if (!mctp_pcie_vdm_wq) {
+		pr_err("Failed to create mctp pcie vdm workqueue\n");
+		return -ENOMEM;
+	}
+
+	return 0;
+}
+
+static __exit void mctp_pcie_vdm_mod_exit(void)
+{
+	struct mctp_pcie_vdm_dev *vdm_dev;
+	struct mctp_pcie_vdm_dev *tmp;
+
+	list_for_each_entry_safe(vdm_dev, tmp, &mctp_pcie_vdm_devs, list) {
+		mctp_pcie_vdm_remove_dev(vdm_dev);
+	}
+}
+
+module_init(mctp_pcie_vdm_mod_init);
+module_exit(mctp_pcie_vdm_mod_exit);
+
+MODULE_DESCRIPTION("MCTP PCIe VDM transport");
+MODULE_LICENSE("GPL");
+MODULE_AUTHOR("YH <yh_chung@aspeedtech.com>");
diff --git a/drivers/soc/aspeed/aspeed-mctp.c b/drivers/soc/aspeed/aspeed-mctp.c
index e3fa5392e48c..1cf28171e037 100644
--- a/drivers/soc/aspeed/aspeed-mctp.c
+++ b/drivers/soc/aspeed/aspeed-mctp.c
@@ -330,7 +330,8 @@ struct aspeed_mctp {
 	struct delayed_work rx_det_dwork;
 	u32 rx_det_period_us;
 #ifdef CONFIG_MCTP_TRANSPORT_PCIE_VDM
-	struct net_device *ndev;
+	/* MCTP PCIe VDM device */
+	struct mctp_pcie_vdm_dev *vdm_dev;
 #endif
 };
 
@@ -791,7 +792,7 @@ static void aspeed_mctp_dispatch_packet(struct aspeed_mctp *priv,
 			wake_up_all(&client->wait_queue);
 		}
 #ifdef CONFIG_MCTP_TRANSPORT_PCIE_VDM
-		mctp_pcie_vdm_receive_packet(priv->ndev);
+		mctp_pcie_vdm_notify_rx(priv->vdm_dev);
 #endif
 		aspeed_mctp_client_put(client);
 	} else {
@@ -988,6 +989,7 @@ static void aspeed_mctp_rx_tasklet(unsigned long data)
 			rx_packet = aspeed_mctp_packet_alloc(GFP_ATOMIC);
 			if (rx_packet) {
 				memcpy(&rx_packet->data, hdr, sizeof(rx_packet->data));
+
 				aspeed_mctp_swap_pcie_vdm_hdr(&rx_packet->data);
 
 				aspeed_mctp_dispatch_packet(priv, rx_packet);
@@ -2095,14 +2097,13 @@ static void aspeed_mctp_pcie_setup(struct aspeed_mctp *priv)
 		if (priv->match_data->need_address_mapping)
 			regmap_update_bits(priv->map, ASPEED_MCTP_EID,
 					   MEMORY_SPACE_MAPPING, BIT(31));
-
-		/* Only set TX MPS since HW will parse RX packet to decide how many bytes to receive
-		 * based on the length field in PCIe VDM header.
-		 */
 		if (priv->match_data->dma_need_64bits_width) {
 			tx_max_payload_size =
 				FIELD_GET(TX_MAX_PAYLOAD_SIZE_MASK,
 					  ilog2(ASPEED_MCTP_MTU >> 6));
+			rx_max_payload_size =
+				FIELD_GET(RX_MAX_PAYLOAD_SIZE_MASK,
+					  (ilog2(ASPEED_MCTP_MTU >> 6)) << RX_MAX_PAYLOAD_SIZE_SHIFT);
 		} else {
 			/*
 			 * In ast2600, tx som and eom will not match expected result.
@@ -2114,6 +2115,9 @@ static void aspeed_mctp_pcie_setup(struct aspeed_mctp *priv)
 			tx_max_payload_size =
 				FIELD_GET(TX_MAX_PAYLOAD_SIZE_MASK,
 						fls(ASPEED_MCTP_MTU >> 6));
+			rx_max_payload_size =
+				FIELD_GET(RX_MAX_PAYLOAD_SIZE_MASK,
+					   (fls(ASPEED_MCTP_MTU >> 6)) << RX_MAX_PAYLOAD_SIZE_SHIFT);
 		}
 
 		regmap_update_bits(priv->map, ASPEED_MCTP_ENGINE_CTRL,
@@ -2510,19 +2514,19 @@ static int aspeed_mctp_probe(struct platform_device *pdev)
 	}
 
 #ifdef CONFIG_MCTP_TRANSPORT_PCIE_VDM
-	struct net_device *ndev;
+	struct mctp_pcie_vdm_dev *vdm_dev;
 	struct mctp_client *client;
 
 	/** use priv's default client to send/receive mctp packets */
 	client = aspeed_mctp_create_client(priv);
 	aspeed_mctp_register_default_handler(client);
 
-	ndev = mctp_pcie_vdm_add_dev(priv->dev, &aspeed_mctp_pcie_vdm_ops);
-	if (IS_ERR(ndev)) {
-		dev_err(priv->dev, "Failed to add mctp pcie vdm device Err %ld\n", PTR_ERR(ndev));
+	vdm_dev = mctp_pcie_vdm_add_dev(priv->dev, &aspeed_mctp_pcie_vdm_ops);
+	if (IS_ERR(vdm_dev)) {
+		dev_err(priv->dev, "Failed to add mctp pcie vdm device Err %ld\n", PTR_ERR(vdm_dev));
 		goto out_drv;
 	}
-	priv->ndev = ndev;
+	priv->vdm_dev = vdm_dev;
 #endif
 
 	ret = aspeed_mctp_dma_init(priv);
@@ -2580,7 +2584,7 @@ static int aspeed_mctp_remove(struct platform_device *pdev)
 	struct aspeed_mctp *priv = platform_get_drvdata(pdev);
 
 #ifdef CONFIG_MCTP_TRANSPORT_PCIE_VDM
-	mctp_pcie_vdm_remove_dev(priv->ndev);
+	mctp_pcie_vdm_remove_dev(priv->vdm_dev);
 #endif
 
 	platform_device_unregister(priv->peci_mctp);
diff --git a/include/linux/aspeed-mctp.h b/include/linux/aspeed-mctp.h
index 503b11aff9ff..48628d2abfe2 100644
--- a/include/linux/aspeed-mctp.h
+++ b/include/linux/aspeed-mctp.h
@@ -29,7 +29,7 @@ struct mctp_protocol_hdr {
 } __packed;
 
 #define PCIE_VDM_HDR_SIZE 16
-#define MCTP_BTU_SIZE 64
+#define MCTP_BTU_SIZE 256
 /* The MTU of the ASPEED MCTP can be 64/128/256 */
 #define ASPEED_MCTP_MTU MCTP_BTU_SIZE
 #define PCIE_VDM_DATA_SIZE_DW (ASPEED_MCTP_MTU / 4)
diff --git a/include/linux/mctp-pcie-vdm.h b/include/linux/mctp-pcie-vdm.h
index 64baa3c838c0..da3fad7929b2 100644
--- a/include/linux/mctp-pcie-vdm.h
+++ b/include/linux/mctp-pcie-vdm.h
@@ -11,6 +11,8 @@
 #include <linux/device.h>
 #include <linux/notifier.h>
 
+struct mctp_pcie_vdm_dev;
+
 /**
  * @send_packet: referenced to send packets with PCIe VDM header packed.
  * @recv_packet: referenced multiple times until no RX packet to be handled.
@@ -25,9 +27,16 @@ struct mctp_pcie_vdm_ops {
 	void (*uninit)(struct device *dev);
 };
 
-struct net_device *mctp_pcie_vdm_add_dev(struct device *dev,
-					 const struct mctp_pcie_vdm_ops *ops);
-void mctp_pcie_vdm_receive_packet(struct net_device *ndev);
-void mctp_pcie_vdm_remove_dev(struct net_device *ndev);
+struct mctp_pcie_vdm_dev *mctp_pcie_vdm_add_dev(struct device *dev,
+						const struct mctp_pcie_vdm_ops *ops);
+void mctp_pcie_vdm_remove_dev(struct mctp_pcie_vdm_dev *vdm_dev);
+
+/**
+ * Notify mctp-pcie-vdm that packets are received and ready to be processed.
+ * After notified, mctp-pcie-vdm will call recv_packet() multiple times
+ * until no more packets to be processed. The lower layer driver can keep receiving
+ * packets while the upper layer is processing the received packets.
+ */
+void mctp_pcie_vdm_notify_rx(struct mctp_pcie_vdm_dev *vdm_dev);
 
 #endif	 /* __LINUX_MCTP_PCIE_VDM_H */
diff --git a/net/mctp/route.c b/net/mctp/route.c
index cf5ae2adf1df..b6befb1a2c0a 100644
--- a/net/mctp/route.c
+++ b/net/mctp/route.c
@@ -16,6 +16,8 @@
 #include <linux/netdevice.h>
 #include <linux/rtnetlink.h>
 #include <linux/skbuff.h>
+#include <linux/dynamic_debug.h>
+#include <linux/printk.h>
 
 #include <uapi/linux/if_arp.h>
 
@@ -1174,8 +1176,15 @@ static int mctp_pkttype_receive(struct sk_buff *skb, struct net_device *dev,
 	if (!rt && mh->dest == MCTP_ADDR_NULL && skb->pkt_type == PACKET_HOST)
 		rt = mctp_route_lookup_null(net, dev);
 
-	if (!rt)
+	if (!rt && mh->dest == MCTP_ADDR_ANY && skb->pkt_type == PACKET_HOST) {
+		pr_debug("%s: Broadcasting", __func__);
+		rt = mctp_route_lookup_null(net, dev);
+	}
+
+	if (!rt) {
+		pr_err("%s: Drop the packet due to no route.", __func__);
 		goto err_drop;
+	}
 
 	rt->output(rt, skb);
 	mctp_route_release(rt);
-- 
2.45.2

